include/boost/gil/image_processing/diffusion.hpp

```diff
@@ -0,0 +1,361 @@
+//
+// Copyright 2020 Olzhas Zhumabek <anonymous.from.applecity@gmail.com>
+//
+// Use, modification and distribution are subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+//
+#include <boost/gil/algorithm.hpp>
+#include <boost/gil/image.hpp>
+#include <boost/gil/image_view.hpp>
+#include <boost/gil/pixel.hpp>
+#include <boost/gil/typedefs.hpp>
+
+#include <vector>
+
+namespace boost { namespace gil {
+namespace detail {
+namespace direction {
+enum
+{
+    north = 0,
+    south = 1,
+    west = 2,
+    east = 3,
+    north_east = 4,
+    south_east = 5,
+    south_west = 6,
+    north_west = 7
+};
+}
+
+// nabla is directed difference in pixels,
+// nabla north for example represents intensity
+// change in the north direction (up), and nabla south
+// represents intensity change in south direction (down).
+template <typename InputView, typename OutputView>
+void compute_nabla(InputView view, std::vector<OutputView> const &nabla)
+{
+    constexpr std::ptrdiff_t input_num_channels = num_channels<InputView>{};
+    using pixel_type = typename OutputView::value_type;
+    static_assert(num_channels<OutputView>{} == input_num_channels,
+                  "input and output views must have the same amount of channels");
+    const auto width = view.width();
+    const auto height = view.height();
+    using namespace direction;
+    const auto zero_pixel = []() {
+        pixel_type pixel;
+        for (std::ptrdiff_t i = 0; i < input_num_channels; ++i)
+        {
+            pixel[i] = 0;
+        }
+
+        return pixel;
+    }();
+    auto minus = [](pixel_type lhs, pixel_type rhs) {
+        for (std::ptrdiff_t i = 0; i < input_num_channels; ++i)
```

Comment:
This lambda needs to capture `input_num_channels` too

Reply:
I'm resolving the conversation because the new version uses different approach.